worksheet_D/oxo.py

```diff
@@ -0,0 +1,176 @@
+
+
+class OxoBoard:
+    def __init__(self):
+        """ The initialiser. Initialise any fields you need here. """
+        self.board = {}
+        for x in xrange(BOARD_EDGE_LENGTH):
+            for y in xrange(BOARD_EDGE_LENGTH):
+                self.board.update({(x, y) : 0})
+
+    def get_square(self, x, y):
+        """ Return 0, 1 or 2 depending on the contents of the specified square. """
+        return self.board.get((x, y))
+
+    def set_square(self, x, y, mark):
+        """ If the specified square is currently empty (0), fill it with mark and return True.
+            If the square is not empty, leave it as-is and return False. """
+        if self.get_square(x, y) == 0:
+            self.board[(x, y)] = mark
+            return True
+        else:
+            return False
+
+    def is_board_full(self):
+        """ If there are still empty squares on the board, return False.
+            If there are no empty squares, return True. """
+
+        # Couldn't get this next line to work. Could it work?
+
+        # Missed the return! Cheers Ed!
+
+        return all(value != 0 for value in self.board.values())
+
+    def get_winner(self):
+        """Revised get_winner function that looks for winning rows of length WINNING_ROW_LENGTH.
+        Clashes with oxotest.py because of the assumed 3x3 grid though.
+        Previous version kept below"""
+        for x in xrange (BOARD_EDGE_LENGTH):
+            for y in xrange (BOARD_EDGE_LENGTH):
+
+                current_contents = self.get_square(x, y)
+
+                if current_contents != 0:
+
+                    # Check for vertical in a row
+                    for i in xrange (WINNING_ROW_LENGTH):
+                        if current_contents != self.get_square(x, y + i):
+                            break
+                    else:
+                        return current_contents
+
+                    # Check for three in a row
+                    for i in xrange (WINNING_ROW_LENGTH):
+                        if current_contents != self.get_square(x + i, y):
+                            break
+                    else:
+                        return current_contents
+
+                    # Check for diagonal down right in a row
+                    for i in xrange (WINNING_ROW_LENGTH):
+                        if current_contents != self.get_square(x + i, y + i):
+                            break
+                    else:
+                        return current_contents
+
+                    #  Check for diagonal down left in a row
+                    for i in xrange (WINNING_ROW_LENGTH):
+                        if current_contents != self.get_square(x + i, y - i):
+                            break
+                    else:
+                        return current_contents
+        return 0
+
+    def get_winner_old(self):
+        """ Not used now. Kept it for nostalgia.
```

Comment:
Best to delete old code -- if you're using version control properly, you can always resurrect it later